File: Lib/Telerik/Examples/QuickStart/ChartView/NullValues/ExampleData.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace Telerik.Examples.WinControls.ChartView.NullValues
	Class ExampleData
        Private _data As List(Of FootballTeam)

        Public Property AccumulatedData() As List(Of FootballTeam)
            Get
                Return Me._data
            End Get
            Set(value As List(Of FootballTeam))
                Me._data = value
            End Set
        End Property

        Public Sub New()
            Me.GetData("..\ChartView\NullValues\Premierleaguedata.csv")
        End Sub

        Private Sub GetData(fileName As String)
            Using fileReader As New StreamReader(fileName)
                Me.GetDataCompleted(Me.ParseData(fileReader))
            End Using
        End Sub

        Protected Sub GetDataCompleted(data As IEnumerable)
            Dim footballTeamStats As List(Of FootballTeamStats) = TryCast(data, List(Of FootballTeamStats))

            Dim teamStats As New Dictionary(Of String, FootballTeam)()

            For Each entry As FootballTeamStats In footballTeamStats
                If Not teamStats.ContainsKey(entry.Name) Then
                    Dim team As New FootballTeam()
                    team.Name = entry.Name
                    team.LogoPath = entry.LogoPath
                    team.Stats = New List(Of FootballTeamStats)()
                    team.TotalWins = If(entry.Wins Is Nothing, 0, entry.Wins)
                    team.TotalDraws = If(entry.Draws Is Nothing, 0, entry.Draws)
                    team.TotalLosses = If(entry.Losses Is Nothing, 0, entry.Losses)
                    teamStats.Add(entry.Name, team)
                Else
                    Dim team As FootballTeam = teamStats(entry.Name)
                    team.Stats.Add(entry)
                    team.TotalWins += If(entry.Wins Is Nothing, 0, entry.Wins)
                    team.TotalDraws += If(entry.Draws Is Nothing, 0, entry.Draws)
                    team.TotalLosses += If(entry.Losses Is Nothing, 0, entry.Losses)
                End If
            Next

            Dim items As New List(Of FootballTeam)()
            items.AddRange(teamStats.Values)
            _data = items
        End Sub

        Protected Function ParseData(dataReader As TextReader) As IEnumerable(Of FootballTeamStats)
            Dim line As String
            Dim dataList As New List(Of FootballTeamStats)()

            line = dataReader.ReadLine()

            While line IsNot Nothing

                If String.IsNullOrEmpty(line) Then
                    Continue While
                End If

                Dim data As String() = line.Split(","c)

                Dim stat As New FootballTeamStats()
                stat.Name = data(0).Trim()
                stat.Season = data(1).Trim()
                If Not String.IsNullOrEmpty(data(2)) Then
                    stat.Wins = Integer.Parse(data(2), CultureInfo.InvariantCulture)
                End If
                If Not String.IsNullOrEmpty(data(3)) Then
                    stat.Draws = Integer.Parse(data(3), CultureInfo.InvariantCulture)
                End If
                If Not String.IsNullOrEmpty(data(4)) Then
                    stat.Losses = Integer.Parse(data(4), CultureInfo.InvariantCulture)
                End If
                If Not String.IsNullOrEmpty(data(5)) Then
                    stat.GoalDifference = Integer.Parse(data(5), CultureInfo.InvariantCulture)
                End If
                If Not String.IsNullOrEmpty(data(6)) Then
                    stat.Points = Integer.Parse(data(6), CultureInfo.InvariantCulture)
                End If
                If Not String.IsNullOrEmpty(data(7)) Then
                    stat.Position = Integer.Parse(data(7), CultureInfo.InvariantCulture)
                End If
                stat.LogoPath = String.Format("..\Resources\SoccerTeamsLogos\{0}", data(8).Trim())

                dataList.Add(stat)

                line = dataReader.ReadLine()
            End While

            Return dataList
        End Function
    End Class

End Namespace

File: Lib/Telerik/Examples/QuickStart/ChartView/NullValues/ExampleData.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;

namespace Telerik.Examples.WinControls.ChartView.NullValues
{
    class ExampleData
    {
        List<FootballTeam> _data;

        public List<FootballTeam> AccumulatedData
        {
            get
            {
                return this._data;
            }
            set
            {
                this._data = value;
            }
        }

        public ExampleData()
        {
            this.GetData(@"..\ChartView\NullValues\Premierleaguedata.csv");
        }

        void GetData(string fileName)
        {
            using (StreamReader fileReader = new StreamReader(fileName))
            {
                this.GetDataCompleted(this.ParseData(fileReader));
            }
        }

        protected void GetDataCompleted(IEnumerable data)
        {
            List<FootballTeamStats> footballTeamStats = data as List<FootballTeamStats>;

            Dictionary<string, FootballTeam> teamStats = new Dictionary<string, FootballTeam>();

            foreach (FootballTeamStats entry in footballTeamStats)
            {
                if (!teamStats.ContainsKey(entry.Name))
                {
                    FootballTeam team = new FootballTeam();
                    team.Name = entry.Name;
                    team.LogoPath = entry.LogoPath;
                    team.Stats = new List<FootballTeamStats>();
                    team.TotalWins = entry.Wins == null ? 0 : entry.Wins;
                    team.TotalDraws = entry.Draws == null ? 0 : entry.Draws;
                    team.TotalLosses = entry.Losses == null ? 0 : entry.Losses;
                    teamStats.Add(entry.Name, team);
                }
                else
                {
                    FootballTeam team = teamStats[entry.Name];
                    team.Stats.Add(entry);
                    team.TotalWins += entry.Wins == null ? 0 : entry.Wins;
                    team.TotalDraws += entry.Draws == null ? 0 : entry.Draws;
                    team.TotalLosses += entry.Losses == null ? 0 : entry.Losses;
                }
            }

            List<FootballTeam> items = new List<FootballTeam>();
            items.AddRange(teamStats.Values);
            _data = items;
        }

        protected IEnumerable<FootballTeamStats> ParseData(TextReader dataReader)
        {
            string line;
            List<FootballTeamStats> dataList = new List<FootballTeamStats>();

            while ((line = dataReader.ReadLine()) != null)
            {
                if (string.IsNullOrEmpty(line))
                    continue;

                string[] data = line.Split(',');

                var stat = new FootballTeamStats();
                stat.Name = data[0].Trim();
                stat.Season = data[1].Trim();
                if (!string.IsNullOrEmpty(data[2]))
                {
                    stat.Wins = int.Parse(data[2], CultureInfo.InvariantCulture);
                }
                if (!string.IsNullOrEmpty(data[3]))
                {
                    stat.Draws = int.Parse(data[3], CultureInfo.InvariantCulture);
                }
                if (!string.IsNullOrEmpty(data[4]))
                {
                    stat.Losses = int.Parse(data[4], CultureInfo.InvariantCulture);
                }
                if (!string.IsNullOrEmpty(data[5]))
                {
                    stat.GoalDifference = int.Parse(data[5], CultureInfo.InvariantCulture);
                }
                if (!string.IsNullOrEmpty(data[6]))
                {
                    stat.Points = int.Parse(data[6], CultureInfo.InvariantCulture);
                }
                if (!string.IsNullOrEmpty(data[7]))
                {
                    stat.Position = int.Parse(data[7], CultureInfo.InvariantCulture);
                }
                stat.LogoPath = string.Format("..\\Resources\\SoccerTeamsLogos\\{0}", data[8].Trim());

                dataList.Add(stat);
            }

            return dataList;
        }
    }
}
